### ISSUE TITLE
delete catalog rule sometimes leave items in catalogrule_product

### ISSUE THREAD

**Issue by peterlembke** [magento/magento2#19278]
### Preconditions (*)
Magento 2.1.10, busy site with a lot of work done on the > 150 catalog rules. Add/Delete/Modify.
Magento 2.2.x - Checked some sites and there are no constraints there either.

### Steps to reproduce (*)
Run
```
select distinct rule_id from catalogrule_product where rule_id NOT IN (select rule_id from catalogrule)
```
and see if you have any lingering items in catalogrule_product

### Expected result (*)
There should be no items left in catalogrule_product that are associated with a deleted catalogrule.

### Actual result (*)
The table catalogrule_product might have some items left.

### Why is this possible
In table catalogrule you should be able to delete a row in the table and then all the associated tables will remove the associations to that rule. Most tables do that: catalogrule_customer_group, catalogrule_group_website, catalogrule_website. So cascading delete is set up right on almost all tables involved.

One table that do not do this are: catalogrule_product. Here you find the product associations to a rule. When a rule are deleted from the table: "catalogrule" then the associated items in table: "catalogrule_product" should also be deleted, but they are not.

Instead this deletion is done in code somewhere. And sometimes that code fails to clean up the table.

### Why does lingering items affect the system?
If the table "catalogrule_product" contain items that are associated with a rule_id that do no longer exist then we get issues when the catalogrule indexer run this insert:
```
INSERT INTO `catalogrule_group_website` (`rule_id`, `customer_group_id`, `website_id`)
SELECT DISTINCT  `catalogrule_product`.`rule_id`, `catalogrule_product`.`customer_group_id`, `catalogrule_product`.`website_id` FROM `catalogrule_product` WHERE (1542708810 >= from_time AND ((1542708810 <= to_time AND to_time > 0) OR to_time = 0))
ON DUPLICATE KEY UPDATE `rule_id` = VALUES(`rule_id`), `customer_group_id` = VALUES(`customer_group_id`), `website_id` = VALUES(`website_id`)
```
The table catalogrule_group_website has a constraint and will only allow rule_ids that exist. There will be a constraint error when catalogrule_product try to insert a none existing rule_id.

### Does the indexer handle this?
The indexer
```
magento indexer:reindex catalogrule_product
```
Will update the items in table catalogrule_product only for item rules that exist. The rest of the items will remain there.

### Proposed solution
Start using constraints more, just as it was done in Magento 1.6.0.0 and older versions.
Delete the lingering items:
```
delete from catalogrule_product where rule_id NOT IN (select rule_id from catalogrule)
```
and add constraints to the table. There are four fields that reference IDs to other tables so here is a constraint on each of these fields:
```
ALTER TABLE `catalogrule_product` ADD CONSTRAINT `CATALOGRULE_PRODUCT_PRODUCT_ID_CATALOG_PRODUCT_ENTITY_ENTITY_ID` FOREIGN KEY (`product_id`) REFERENCES `catalog_product_entity` (`entity_id`) ON DELETE CASCADE ON UPDATE CASCADE;

ALTER TABLE `catalogrule_product` ADD CONSTRAINT `CATALOGRULE_PRODUCT_CUSTOMER_GRP_ID_CUSTOMER_GRP_CUSTOMER_GRP_ID` FOREIGN KEY (`customer_group_id`) REFERENCES `customer_group` (`customer_group_id`) ON DELETE CASCADE ON UPDATE CASCADE;

ALTER TABLE `catalogrule_product` ADD CONSTRAINT `CATALOGRULE_PRODUCT_RULE_ID_CATALOGRULE_RULE_ID` FOREIGN KEY (`rule_id`) REFERENCES `catalogrule` (`rule_id`) ON DELETE CASCADE ON UPDATE CASCADE;

ALTER TABLE `catalogrule_product` ADD CONSTRAINT `CATALOGRULE_PRODUCT_WEBSITE_ID_STORE_WEBSITE_WEBSITE_ID` FOREIGN KEY (`website_id`) REFERENCES `store_website` (`website_id`) ON DELETE CASCADE ON UPDATE CASCADE;
```
After this is applied I have tested that catalog rules still work.

**Comment by magento-engcom-team** [comment]
Hi @peterlembke. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [ ] Summary of the issue
- [ ] Information on your environment
- [ ] Steps to reproduce
- [ ] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me $VERSION instance` 

where `$VERSION` is [version tags](https://github.com/magento/magento2/tags) (starting from 2.2.0+) or develop branches (for example: 2.3-develop). 
For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@peterlembke do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [ ] no

**Comment by orlangur** [comment]
I believe such constraints were removed intentionally, thus returning them back is not a viable solution. Need to be found from Git history.

**Comment by tandih** [comment]
Hi, 

Face this issue today on our project and i wanna to say thank you for your proposed solution @peterlembke

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-nazar. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [x] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [x] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [x] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [x] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [x] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [x] 6. Add label `Issue: Confirmed` once verification is complete. 

- [x] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by magento-engcom-team** [comment]
@engcom-backlog-nazar Thank you for verifying the issue. Based on the provided information internal tickets `MAGETWO-96654`, `MAGETWO-96655` were created

**Comment by magento-engcom-team** [comment]
@peterlembke, thank you for your report.     
    
Unfortunately, we are archiving this ticket now as it did not get much attention from both Magento Community and Core developers for an extended period. This is done in an effort to create a quality, community-driven backlog which will allow us to allocate the required attention more easily.

Please feel free to comment, reopen or create new ticket according to the [Issue reporting guidelines](https://github.com/magento/magento2/wiki/Issue-reporting-guidelines) 
if you are still facing this issue on the latest `2.3-develop` branch. Thank you for collaboration.

**Comment by tuyennn** [comment]
@peterlembke @magento-engcom-team I have same issue experience on Magento 2.1.18, probably will have update to 2.3 but not now.
